homework2/task4.py

```diff
@@ -0,0 +1,35 @@
+"""
+Write a function that accepts another function as an argument. Then it
+should return such a function, so the every call to initial one
+should be cached.
+def func(a, b):
+    return (a ** b) ** 2
+
+cache_func = cache(func)
+some = 100, 200
+val_1 = cache_func(*some)
+val_2 = cache_func(*some)
+assert val_1 is val_2
+"""
+from collections import defaultdict
+from typing import Callable
+
+
+def cache(func: Callable) -> Callable:
+    """Cashe every call to initial function.
+    (decorator).
+
+    Args:
+        input function
+    Returns:
+        such a function, so the every call to initial one
+    should be cached.
+    """
+    cache_dict = defaultdict(dict)
+
+    def a_wrapper(*args, **kwargs):
+        if args not in cache_dict:
+            cache_dict[args] = func(*args, **kwargs)
```

Comment:
I have 2 concerns here:
1. Could you change the function so that kwargs are also considered in caching logic? We can have the calls like these: func(1,2, c=4) and func(1,2, c=6). And they are supposed to give different results, but according to the current logic these are the same function calls.
2. What will happen if you pass mutable data types in args/kwargs? Change the logic so that keys in `cache_dict `are immutable.

Reply:
fixed